src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -2858,95 +2858,436 @@ private static CommandBaseAst GetPreviousPipelineCommand(CommandAst commandAst)
+                    varName = userPath;
+                }
+                else
+                {
+                    scopeName = userPath.Remove(scopeIndex);
+                    varName = userPath.Substring(scopeIndex + 1);
+                }
+
+                if (!varName.EqualsOrdinalIgnoreCase(VariableTarget.VariablePath.UnqualifiedPath))
+                {
+                    return false;
                 }
+
+                return AssignsToTargetScope(scopeName);
             }
 
-            return res;
-        }
+            private bool AssignsToTargetScope(string scopeName)
+                => LocalScopeOnly
+                    ? string.IsNullOrEmpty(scopeName) || scopeName.EqualsOrdinalIgnoreCase("Local") || scopeName.EqualsOrdinalIgnoreCase("Private")
+                    : ScopeIsLocal || !(scopeName.EqualsOrdinalIgnoreCase("Local") || scopeName.EqualsOrdinalIgnoreCase("Private"));
 
-        public static bool AstAssignsToSameVariable(this VariableExpressionAst variableAst, Ast ast)
-        {
-            var parameterAst = ast as ParameterAst;
-            var variableAstVariablePath = variableAst.VariablePath;
-            if (parameterAst != null)
+            public override AstVisitAction DefaultVisit(Ast ast)
             {
-                return variableAstVariablePath.IsUnscopedVariable &&
-                       parameterAst.Name.VariablePath.UnqualifiedPath.Equals(variableAstVariablePath.UnqualifiedPath, StringComparison.OrdinalIgnoreCase) &&
-                       parameterAst.Parent.Parent.Extent.EndOffset > variableAst.Extent.StartOffset;
+                if (ast.Extent.StartOffset >= StopSearchOffset)
+                {
+                    // When visiting do while/until statements, the condition will be visited before the statement block.
+                    // The condition itself may not be interesting if it's after the cursor, but the statement block could be.
+                    return ast is PipelineBaseAst && ast.Parent is DoUntilStatementAst or DoWhileStatementAst
+                        ? AstVisitAction.SkipChildren
+                        : AstVisitAction.StopVisit;
+                }
+
+                return AstVisitAction.Continue;
             }
 
-            if (ast is ForEachStatementAst foreachAst)
+            public override AstVisitAction VisitAssignmentStatement(AssignmentStatementAst assignmentStatementAst)
             {
-                return variableAstVariablePath.IsUnscopedVariable &&
-                       foreachAst.Variable.VariablePath.UnqualifiedPath.Equals(variableAstVariablePath.UnqualifiedPath, StringComparison.OrdinalIgnoreCase);
+                if (assignmentStatementAst.Extent.StartOffset >= StopSearchOffset)
+                {
+                    return assignmentStatementAst.Parent is DoUntilStatementAst or DoWhileStatementAst
+                        ? AstVisitAction.SkipChildren
+                        : AstVisitAction.StopVisit;
```

Comment:
The same question.

And could you please show script examples we process here?